src/prommis/nanofiltration/nf_brine.py

```diff
@@ -0,0 +1,386 @@
+    m.fs.permeate = Product(property_package=m.fs.properties)
+    m.fs.retentate = Product(property_package=m.fs.properties)
+
+    # define unit models
+    m.fs.pump = Pump(property_package=m.fs.properties)
+    m.fs.unit = NanofiltrationDSPMDE0D(property_package=m.fs.properties)
+
+    # connect the streams and blocks
+    m.fs.feed_to_pump = Arc(source=m.fs.feed.outlet, destination=m.fs.pump.inlet)
+    m.fs.pump_to_nf = Arc(source=m.fs.pump.outlet, destination=m.fs.unit.inlet)
+    m.fs.nf_to_permeate = Arc(
+        source=m.fs.unit.permeate, destination=m.fs.permeate.inlet
+    )
+    m.fs.nf_to_retentate = Arc(
+        source=m.fs.unit.retentate, destination=m.fs.retentate.inlet
+    )
+    TransformationFactory("network.expand_arcs").apply_to(m)
+    return m
+
+
+def fix_initial_variables(m):
+    """
+    Fixes the initial variables needed to create 0 DOF
+
+    Args:
+        m: pyomo model
+    """
+
+    # pump variables
+    m.fs.pump.efficiency_pump[0].fix(0.75)
+    m.fs.pump.control_volume.properties_in[0].temperature.fix(298.15)
+    m.fs.pump.control_volume.properties_in[0].pressure.fix(101325)
+    m.fs.pump.outlet.pressure[0].fix(2e5)
+    iscale.set_scaling_factor(m.fs.pump.control_volume.work, 1e-4)
+
+    # membrane operation
+    m.fs.unit.recovery_vol_phase[0, "Liq"].setub(0.95)
+    m.fs.unit.spacer_porosity.fix(0.85)
+    m.fs.unit.channel_height.fix(5e-4)
+    m.fs.unit.velocity[0, 0].fix(0.1)
+    m.fs.unit.area.fix(100)
+    m.fs.unit.mixed_permeate[0].pressure.fix(101325)
+
+    # variables for calculating mass transfer coefficient with spiral wound correlation
+    m.fs.unit.spacer_mixing_efficiency.fix()
+    m.fs.unit.spacer_mixing_length.fix()
+
+    # membrane properties
+    m.fs.unit.radius_pore.fix(0.5e-9)
+    m.fs.unit.membrane_thickness_effective.fix(1.33e-6)
+    m.fs.unit.membrane_charge_density.fix(-60)
+    m.fs.unit.dielectric_constant_pore.fix(41.3)
+    iscale.calculate_scaling_factors(m)
+
+
+def unfix_optimization_variables(m):
+    """
+    Unfixes select variables to enable optimization with DOF>0
+
+    Args:
+        m: pyomo model
+    """
+    m.fs.pump.outlet.pressure[0].unfix()
+    m.fs.unit.area.unfix()
+
+
+def add_objective(m):
+    """
+    Adds objectives to the pyomo model
+
+    Args:
+        m: pyomo model
+    """
+    # limit Li loss
+    m.fs.objective = Objective(
+        expr=m.fs.retentate.flow_mol_phase_comp[0, "Liq", "Li_+"]
+    )
+
+
+def add_pressure_constraint(m, pressure_limit=7e6):
+    """
+    Adds feed pressure constraint to the pyomo model
+
+    Args:
+        m: pyomo model
+        pressure_limit: upper bound on the outlet pump pressure
+    """
+    # bound the feed pressure to a reasonable value for nanofiltration
+    # choose an upper limit of 70 bar (https://doi.org/10.1021/acs.est.2c08584)
+    m.fs.pressure_constraint = Constraint(
+        expr=m.fs.pump.outlet.pressure[0] <= pressure_limit
+    )
+
+
+def optimize(m, solver):
```

Comment:
I think this name is slightly misleading - this is solving the model rather than strictly optimizing (i.e. it also works for deterministic problems).